src/ast/local-graph.h

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_ast_local_graph_h
+#define wasm_ast_local_graph_h
+
+#include <iterator>
+
+#include <wasm-builder.h>
+#include <ast/find_all.h>
+
+namespace wasm {
+
+//
+// Finds the connections between get_locals and set_locals, creating
+// a graph of those ties. This is useful for "ssa-style" optimization,
+// in which you want to know exactly which sets are relevant for a
+// a get, so it is as if each get has just one set, logically speaking
+// (see the SSA pass for actually creating new local indexes based
+// on this).
+//
+// TODO: the algorithm here is pretty simple, but also pretty slow,
+//       we should optimize it. e.g. we rely on set_interaction
+//       here, and worse we only use it to compute the size...
+struct LocalGraph : public PostWalker<LocalGraph> {
```

Comment:
Can we put this in a cpp file rather than a header file? My build times thank you for it :)

Reply:
Sure, done. Even if I kind of think it's more elegant the other way... ;)

I wasn't sure about the file name. We seem to use FileName.cpp but file-name.h for other things...